[Sema] Check that return and argument types of a virtual member function

ahatanaka opened the pull request:
are complete if the address of the member function is taken

If they aren't complete, reject the code.

rdar://problem/60888064

ahatanaka committed:
[Sema] Check that return and argument types of a virtual member function
are complete if the address of the member function is taken

If they aren't complete, reject the code.

rdar://problem/60888064

jfbastien commented:
It would be nice to explain why the type needs to be complete, as a comment in the code.

FWIW this constrains the language, but I'm not sure we care about this specific instance.

I'd like to address this general issue with this: rdar://problem/60436372
But it'll take a while to get there so IMO this is a fine restriction.

rjmccall commented:
I pinged Richard about whether this is legal as a general matter.  I think it should be, thunking is intended to be an acceptable implementation strategy for virtual member function pointers as far as I can tell.

This can go upstream, I think.

rjmccall commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -12681,6 +12681,40 @@ QualType Sema::CheckAddressOfOperand(ExprResult &OrigOp, SourceLocation OpLoc) {
 
+    if (getASTContext().getTargetInfo().getTriple().getArch() ==
+            llvm::Triple::aarch64 &&
+        getLangOpts().PointerAuthCalls && MD->isVirtual()) {
```
Checking PointerAuthCalls is fine, no need for the target check.  Assuming we don't want to check this more generally.  Do we not have a check for this for MSVC?

ahatanaka replied:
clang doesn't have any problems compiling the code in the test case when the target is MSVC (i386-pc-win32) since it can emit variadic thunks I mentioned in the radar. If we are going to upstream this, should this be a warning instead of an error for targets that aren't arm64e? I don't think we can start rejecting code that currently works fine?

rjmccall replied:
We've done so before when it was clearly a bug that it was accepted.  I think we'd need some committee guidance about it, though.

ahatanaka replied:
OK, I'll send a patch or an RFC to the list later.

rjmccall replied:
Don't worry about it, I wrote up a DR for the C++ committee.

rjmccall commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -12681,6 +12681,40 @@ QualType Sema::CheckAddressOfOperand(ExprResult &OrigOp, SourceLocation OpLoc) {
+      if (IsIncomplete)
+        return QualType();
+    }
+
```
We might need to trigger instantiation of the `noexcept` clause.

ahatanaka replied:
Sorry, I didn't understand this. Do you mean there should be a test for a virtual function which has the `noexcept` specifier? I'm not sure how that changes the diagnostic.

rjmccall replied:
If there's a class template with a virtual function with a dependent `noexcept` clause, and we form a member pointer to that virtual function, we should make sure we don't have problems, which probably means making sure we instantiate the clause.

ahatanaka replied:
The check `MPTy->isDependentType` isn't needed to emit the diagnostics in the test case, but I added it since it wasn't clear whether there is a guarantee that `Sema::CheckAddressOfOperand` won't be called when the type is dependent.

rjmccall reviewed (commented):
A few comments on the patch as-is, but I think we should try to upstream this; I think this is an oversight in the spec.

ahatanaka committed:
Address review comments

Explain why the types must be complete when pointer authentication is
enabled. Check that virual function pointers aren't diagnosed in
unevaluated contexts. Change the variable and diagnostic names to
improve indentation.

ahatanaka committed:
Check that a pointer to a virtual function pointer with a dependent
noexcept clause isn't diagnosed

Test virtual functions of templated classes.

rjmccall reviewed (approved):
Okay, we can take in our branch for now.
